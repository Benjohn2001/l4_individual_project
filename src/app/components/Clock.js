import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { TouchableOpacity, Text,  View, Vibration } from "react-native";
import { Feather } from '@expo/vector-icons';
import { getStorage, getDownloadURL, uploadBytesResumable, uploadBytes, uploadString } from "firebase/storage";
import {ref as stref} from "firebase/storage";
import { Dimensions } from 'react-native';
import Svg, { Circle, Defs, Pattern, Image, ClipPath } from 'react-native-svg';
import Labels from './Labels';

function Clock (props){

    const {locations, face} = props
    const windowWidth = Dimensions.get("window").width

    const [pic, setPic] = React.useState("")

    const diameter = windowWidth-40
    const center=windowWidth/2
    const radius = diameter/2

    if(face.length===7){
        return(
            <Svg height={windowWidth} width={windowWidth}>
            <Circle cx={center} cy={center} r={radius+15} stroke={"black"} strokeWidth={2} fill="#6B4EFF"/>
            <Circle cx={center} cy={center} r={radius-15} stroke={"black"} strokeWidth={2} fill={face}/>
            <Circle cx={center} cy={center} r={radius-15} stroke={"black"} strokeWidth={2}/>
            <Labels
                radius={radius}
                center={center}
                locations={locations}
            />
        </Svg>
        )
    }else{
        getDownloadURL(stref(getStorage(), face)).then((url)=>{
            setPic(url)
        })
    return(
        <Svg height={windowWidth} width={windowWidth}>

            <Defs>
                <ClipPath id="clip" >
                    <Circle cx={center} cy={center} r={radius-15} stroke={"black"} strokeWidth={2} fill="url(#image)"/>
                </ClipPath>
            </Defs>
            <Circle cx={center} cy={center} r={radius+15} stroke={"black"} strokeWidth={2} fill="#6B4EFF"/>
            
            <Image width={windowWidth} height={windowWidth} preserveAspectRatio="none" href={{uri: pic}} clipPath="url(#clip)"/>
            <Circle cx={center} cy={center} r={radius-15} stroke={"black"} strokeWidth={2}/>

            <Labels
                radius={radius}
                center={center}
                locations={locations}
            />
        </Svg>

    ); 
    }
}

Clock.propTypes = {
    locations: PropTypes.array.isRequired,
    face: PropTypes.string.isRequired
}

export default Clock