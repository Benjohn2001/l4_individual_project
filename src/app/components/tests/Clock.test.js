import { render } from '@testing-library/react-native'
import renderer from "react-test-renderer";
import Clock from '../Clock';

test('Check Clock working correctly',()=>{
      
    const data1 = {
        "colour": "#00afef",
        "member": "KKSw9RsMRlSBKjUy7EGOiA1NcFk2",
        "status": 5
    }
    const snap1 = { val: () => data1 };
    const snaps=[snap1]
    const clock = render(<Clock 
        locations={['a','a','a','a','a','a']}
        face="#00afef"
        membs={snaps}
    />)

    const domTree = renderer.create(
        <Clock 
            locations={['a','a','a','a','a','a']}
            face="#00afef"
            membs={snaps}
        />
        ).toJSON();
      expect(domTree).toMatchSnapshot();

    expect(clock).toBeTruthy()
})



