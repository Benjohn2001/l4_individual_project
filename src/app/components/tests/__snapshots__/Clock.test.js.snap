// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check Clock working correctly 1`] = `
<RNSVGSvgView
  bbHeight={750}
  bbWidth={750}
  focusable={false}
  height={750}
  style={
    Array [
      Object {
        "backgroundColor": "transparent",
        "borderWidth": 0,
      },
      Object {
        "flex": 0,
        "height": 750,
        "width": 750,
      },
    ]
  }
  width={750}
>
  <RNSVGGroup>
    <RNSVGCircle
      cx={375}
      cy={375}
      fill={4285222655}
      propList={
        Array [
          "fill",
          "stroke",
          "strokeWidth",
        ]
      }
      r={370}
      stroke={4278190080}
      strokeWidth={2}
    />
    <RNSVGCircle
      cx={375}
      cy={375}
      fill={4278235119}
      propList={
        Array [
          "fill",
          "stroke",
          "strokeWidth",
        ]
      }
      r={340}
      stroke={4278190080}
      strokeWidth={2}
    />
    <RNSVGCircle
      cx={375}
      cy={375}
      propList={
        Array [
          "stroke",
          "strokeWidth",
        ]
      }
      r={340}
      stroke={4278190080}
      strokeWidth={2}
    />
    <RNSVGGroup>
      <RNSVGGroup>
        <RNSVGLine
          propList={
            Array [
              "stroke",
              "strokeWidth",
              "strokeLinecap",
            ]
          }
          stroke={4278190080}
          strokeLinecap={1}
          strokeWidth={2}
          x1={375}
          x2={375}
          y1={375}
          y2={5}
        />
        <RNSVGText
          alignmentBaseline="central"
          content={null}
          dx={Array []}
          dy={Array []}
          fill={4278190080}
          font={
            Object {
              "fontSize": 15,
              "fontWeight": "bold",
              "textAnchor": "middle",
            }
          }
          matrix={
            Array [
              0.8660254037844387,
              0.49999999999999994,
              -0.49999999999999994,
              0.8660254037844387,
              552.5,
              67.56098165652429,
            ]
          }
          propList={
            Array [
              "fill",
            ]
          }
          rotate={Array []}
          x={Array []}
          y={Array []}
        >
          <RNSVGTSpan
            content="a"
            dx={Array []}
            dy={Array []}
            font={Object {}}
            rotate={Array []}
            x={Array []}
            y={Array []}
          />
        </RNSVGText>
      </RNSVGGroup>
      <RNSVGGroup>
        <RNSVGLine
          propList={
            Array [
              "stroke",
              "strokeWidth",
              "strokeLinecap",
            ]
          }
          stroke={4278190080}
          strokeLinecap={1}
          strokeWidth={2}
          x1={375}
          x2={695.4293994002423}
          y1={375}
          y2={190.00000000000003}
        />
        <RNSVGText
          alignmentBaseline="central"
          content={null}
          dx={Array []}
          dy={Array []}
          fill={4278190080}
          font={
            Object {
              "fontSize": 15,
              "fontWeight": "bold",
              "textAnchor": "middle",
            }
          }
          matrix={
            Array [
              6.123233995736766e-17,
              1,
              -1,
              6.123233995736766e-17,
              730,
              375,
            ]
          }
          propList={
            Array [
              "fill",
            ]
          }
          rotate={Array []}
          x={Array []}
          y={Array []}
        >
          <RNSVGTSpan
            content="a"
            dx={Array []}
            dy={Array []}
            font={Object {}}
            rotate={Array []}
            x={Array []}
            y={Array []}
          />
        </RNSVGText>
      </RNSVGGroup>
      <RNSVGGroup>
        <RNSVGLine
          propList={
            Array [
              "stroke",
              "strokeWidth",
              "strokeLinecap",
            ]
          }
          stroke={4278190080}
          strokeLinecap={1}
          strokeWidth={2}
          x1={375}
          x2={695.4293994002423}
          y1={375}
          y2={560}
        />
        <RNSVGText
          alignmentBaseline="central"
          content={null}
          dx={Array []}
          dy={Array []}
          fill={4278190080}
          font={
            Object {
              "fontSize": 15,
              "fontWeight": "bold",
              "textAnchor": "middle",
            }
          }
          matrix={
            Array [
              0.8660254037844387,
              -0.49999999999999994,
              0.49999999999999994,
              0.8660254037844387,
              552.5,
              682.4390183434757,
            ]
          }
          propList={
            Array [
              "fill",
            ]
          }
          rotate={Array []}
          x={Array []}
          y={Array []}
        >
          <RNSVGTSpan
            content="a"
            dx={Array []}
            dy={Array []}
            font={Object {}}
            rotate={Array []}
            x={Array []}
            y={Array []}
          />
        </RNSVGText>
      </RNSVGGroup>
      <RNSVGGroup>
        <RNSVGLine
          propList={
            Array [
              "stroke",
              "strokeWidth",
              "strokeLinecap",
            ]
          }
          stroke={4278190080}
          strokeLinecap={1}
          strokeWidth={2}
          x1={375}
          x2={375}
          y1={375}
          y2={745}
        />
        <RNSVGText
          alignmentBaseline="central"
          content={null}
          dx={Array []}
          dy={Array []}
          fill={4278190080}
          font={
            Object {
              "fontSize": 15,
              "fontWeight": "bold",
              "textAnchor": "middle",
            }
          }
          matrix={
            Array [
              0.8660254037844387,
              0.49999999999999994,
              -0.49999999999999994,
              0.8660254037844387,
              197.50000000000009,
              682.4390183434757,
            ]
          }
          propList={
            Array [
              "fill",
            ]
          }
          rotate={Array []}
          x={Array []}
          y={Array []}
        >
          <RNSVGTSpan
            content="a"
            dx={Array []}
            dy={Array []}
            font={Object {}}
            rotate={Array []}
            x={Array []}
            y={Array []}
          />
        </RNSVGText>
      </RNSVGGroup>
      <RNSVGGroup>
        <RNSVGLine
          propList={
            Array [
              "stroke",
              "strokeWidth",
              "strokeLinecap",
            ]
          }
          stroke={4278190080}
          strokeLinecap={1}
          strokeWidth={2}
          x1={375}
          x2={54.570600599757654}
          y1={375}
          y2={560}
        />
        <RNSVGText
          alignmentBaseline="central"
          content={null}
          dx={Array []}
          dy={Array []}
          fill={4278190080}
          font={
            Object {
              "fontSize": 15,
              "fontWeight": "bold",
              "textAnchor": "middle",
            }
          }
          matrix={
            Array [
              6.123233995736766e-17,
              -1,
              1,
              6.123233995736766e-17,
              20,
              375.00000000000006,
            ]
          }
          propList={
            Array [
              "fill",
            ]
          }
          rotate={Array []}
          x={Array []}
          y={Array []}
        >
          <RNSVGTSpan
            content="a"
            dx={Array []}
            dy={Array []}
            font={Object {}}
            rotate={Array []}
            x={Array []}
            y={Array []}
          />
        </RNSVGText>
      </RNSVGGroup>
      <RNSVGGroup>
        <RNSVGLine
          propList={
            Array [
              "stroke",
              "strokeWidth",
              "strokeLinecap",
            ]
          }
          stroke={4278190080}
          strokeLinecap={1}
          strokeWidth={2}
          x1={375}
          x2={54.57060059975771}
          y1={375}
          y2={189.99999999999997}
        />
        <RNSVGText
          alignmentBaseline="central"
          content={null}
          dx={Array []}
          dy={Array []}
          fill={4278190080}
          font={
            Object {
              "fontSize": 15,
              "fontWeight": "bold",
              "textAnchor": "middle",
            }
          }
          matrix={
            Array [
              0.8660254037844387,
              -0.49999999999999994,
              0.49999999999999994,
              0.8660254037844387,
              197.49999999999983,
              67.56098165652435,
            ]
          }
          propList={
            Array [
              "fill",
            ]
          }
          rotate={Array []}
          x={Array []}
          y={Array []}
        >
          <RNSVGTSpan
            content="a"
            dx={Array []}
            dy={Array []}
            font={Object {}}
            rotate={Array []}
            x={Array []}
            y={Array []}
          />
        </RNSVGText>
      </RNSVGGroup>
    </RNSVGGroup>
  </RNSVGGroup>
</RNSVGSvgView>
`;
